src/sophia/utility.f

```diff
@@ -0,0 +1,69 @@
+      SUBROUTINE TOEVT()
+C************************************************************************
+C
+C     Converts data in S_PLIST common block to
+C     standard HEPEVT common block format, 
+C     and S_PLIST1 and S_CHP to SCHG common block containing
+C     charge and parent decayed particle information
+C     
+C************************************************************************
+      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
+      IMPLICIT INTEGER (I-N)
+
+      SAVE
+      INTEGER PDGCOD(49)
+      DATA PDGCOD /
+     &  22, -11, 11, -13, 13, 111, 211, -211, 321, -321, 130, 310, 2212,
+     &  2112, 12, -12, 14, -14, -99999999, -99999999, 311, -311, 221, 
+     &  331, 213, -213, 113, 323, -323, 313, -313, 223, 333, 3222, 3212,
+     &  3112, 3322, 3312, 3122, 2224, 2214, 2114, 1114, 3224, 3214, 
+     &  3114, 3324, 3314, 3334 /
+       
+      COMMON /S_PLIST/ P(2000,5), LLIST(2000), NP, Ideb
+      COMMON /S_PLIST1/ LLIST1(2000)
+      COMMON /S_CHP/ S_LIFE(49), ICHP(49), ISTR(49), IBAR(49)
+        
+C     Standard HEPEVT common block:
+      PARAMETER (NMXHEP = 2000)
+      PARAMETER (NMXHP2 = NMXHEP * 2, NMXHP4 = NMXHEP * 4, 
+     & NMXHP5 = NMXHEP * 5)
+      COMMON /HEPEVT/ NEVHEP, NHEP, ISTHEP(NMXHEP), IDHEP(NMXHEP),
+     & JMOHEP(2, NMXHEP), JDAHEP(2, NMXHEP), 
+     & PHEP(5, NMXHEP), VHEP(4, NMXHEP)
+
+C     Initialization of HEVEVT variables
+      DATA NEVHEP / 0 /
+      DATA JMOHEP/ NMXHP2 * 0.0/, JDAHEP / NMXHP2 * 0.0 /
+      DATA PHEP/ NMXHP5 * 0.0/, VHEP / NMXHP4 * 0.0 /
+
+C     Common block SCHG with additional information:
+C     ICHG - charge, 
+C     DECPAR - position in the arrays (ISTHEP, PHEP, ...) of parent (decayed) particle     
+      INTEGER ICHG, DECPAR
+      COMMON /SCHG/ ICHG(NMXHEP), DECPAR(NMXHEP)
+
+      INTEGER CID  ! id of the current (in the loop) particle
+C     number added to decayed particle:
+      INTEGER DNUM
+      PARAMETER (DNUM = 10000)
+       
+      NEVHEP = NEVHEP + 1
+      NHEP = NP
+C     We do not transpose P, because we don't use PHEP         
+C      PHEP = TRANSPOSE(P)
```

Comment:
Not using PHEP is a bit non standard for the rest of impy. Not sure I recommend really doing this because the HEPEVT is a standard common block and could be used as such by someone else with the corresponding interface.

How much penalty in ns,us, ms does it cost to fill PHEP?

In the end this must work with the HepMC3 interface, so whatever it requires. Also, consider that I expect most people to use python to analyse the output. So making numpy computations work on non-contiguous arrays might kill the benefit you gain here on the fortran side. However, I never tested this.

Reply:
The operation PHEP = TRANSPOSE(P) adds approximately 10% to the execution time of bare event generation. Probably, it is still not critical, considering that subsequent operations in python will eat up  much more. I commented it out because it turns out, that the operation like
```python
px, py, pz, en, m = lib.s_plist.p[:np, :].T
```
produces contiguous arrays px, py, pz, en, m. And if we used transposed PHEP matrix, px, py, pz, en, m will not be contiguous in memory. This is because  lib.s_plist.p(2000, 5) in the memory goes like p(1, 1), p(2, 1), p(3, 1) and f2py do not change layout in memory. One can verify that px is contiguous by looking at px.__array_interface__ dictionary. 

So from this point of view we don't need PHEP at all. However, from point of view of interfaces, probably PHEP should be filled out using PHEP = TRANSPOSE(P) after all.

Reply:
As I mentioned, there is no need to transpose the entire array. NEVHEP is usually between 2 and 100. So just a for loop is enough